helpers/subscriptionControllers.js

```diff
@@ -0,0 +1,140 @@
+import { cssEscape } from '../../helpers/dom.js';
+
+export class ProviderController {
+
+	constructor(host, callbacks, options) {
+		this._host = host;
+		this._callbacks = callbacks;
+		this._eventName = options.eventName;
+		this.subscribers = new Map();
+
+		this._handleSubscribe = this._handleSubscribe.bind(this);
+	}
+
+	hostConnected() {
+		if (this._eventName) this._host.addEventListener(this._eventName, this._handleSubscribe);
+	}
+
+	hostDisconnected() {
+		if (this._eventName) this._host.removeEventListener(this._eventName, this._handleSubscribe);
+	}
+
+	subscribe(target) {
+		if (this.subscribers.has(target)) return;
+		this.subscribers.set(target, target);
+		this.updateSubscribers();
+		if (this._callbacks.onSubscribe) this._callbacks.onSubscribe(target);
+	}
+
+	unsubscribe(target) {
+		this.subscribers.delete(target);
+		if (this._callbacks.onUnsubscribe) this._callbacks.onUnsubscribe(target);
+	}
+
+	updateSubscribers() {
+		if (!this.subscribers || this.subscribers.size === 0) return;
+		if (!this._callbacks.updateSubscribers) return;
+
+		// debounce the updates
+		if (this._updateSubscribersRequested) return;
+
+		this._updateSubscribersRequested = true;
+		setTimeout(() => {
+			this._callbacks.updateSubscribers(this.subscribers);
+			this._updateSubscribersRequested = false;
+		}, 0);
+	}
+
+	_handleSubscribe(e) {
+		e.stopPropagation();
+		e.detail.provider = this._host;
+		const target = e.composedPath()[0];
+		this.subscribe(target);
+	}
+}
+
+export class SubscriberController {
+
+	constructor(host, callbacks, options) {
+		this._host = host;
+		this._callbacks = callbacks;
+		this._eventName = options.eventName;
+		this._forPropertyName = options.forProperty;
+		this._controllerId = options.controllerId;
+		this._providers = new Map();
+	}
+
+	// This method is basically offered for convenience
+	// Components using the event method or a forProperty that takes a string will know they only have one provider
+	get provider() {
+		return this._providers.values().next().value;
+	}
+
+	get providers() {
+		return Array.from(this._providers.values());
+	}
+
+	hostConnected() {
+		if (!this._eventName || this._host[this._forPropertyName]) return;
```

Comment:
The weirdest part about splitting this controller into two is this line.  Should I add a callback to ask the consumer whether they actually want us to fire the event (or if they ended up getting the `for` attribute set instead), or do I just send the event?  Sending the event when we don't need to means:
1. It goes off into space and no one actions it, and we don't care
2. A provider does hear it and adds us as a subscriber, but we basically ignore that provider. There is a hypothetical case where the provider is trying to update us and we don't want it to, but I can specifically protect against that in the selection components

Reply:
Hmm, I think this highlights that I didn't really understand what was going on here. I assumed that the code was only needed in the `EventSubscriberController` case, but it looks like additionally if it's using BOTH controllers then it should still not fire the event?

Reply:
It's only needed for the event case, but the selection components are weird.  They basically ask you to provide a `selectionFor` property, and if you don't, they use the event case.  If you do, they use the `ForProperty/Attribute` case.So I only evere need to use one or the other, but when I'm setting up controllers in the `constructor`, I don't yet know if we're going to get a `for` property.  I'm assuming I can't setup controllers (the real Lit ones) in the `connected` callback?

Reply:
I do think the divided works much better, I'll push that change so you can see